### PR TITLE
Vcs3D features for v2.1.1

### PR DESCRIPTION
**Pull request by ThomasMaxwell**
Fixes a number of minor issues- mostly focussed on plot provenance and initialization from scripts and .vt files.

### CODE REVIEW

**Comment by aashish24** [comment]
@ThomasMaxwell is this branch ready?

**Comment by ThomasMaxwell** [comment]
I am finished adding features to this branch.   Savannah has been testing the code- she found a bug yesterday in the Hovmoller plotter and I pushed a fix for that one this morning.   Please feel free to merge this branch into the 2.1.1 release branch and then we will continue our testing on the full release.

-- Tom

**Comment by aashish24** [comment]
I will review and merge it today. Thanks!

**Comment by aashish24** [comment]
@ThomasMaxwell I guess this is ready for review?

**Comment by ThomasMaxwell** [comment]
Yes, you can review this branch.

**Comment by ThomasMaxwell** [comment]
Please don't forget: has corresponding vistrails branch: uvcdat-vcs3D-features_for_v2.1.1.

**Comment by aashish24** [comment]
Sorry I missed it earlier. I will review it soon.

File: .hooks-config [hunk 1]
```diff
@@ -1,5 +0,0 @@
-[hooks "chain"]
```
Comment (aashish24) [inline .hooks-config]: why removed this?

**Comment by aashish24** [comment]
@dlonie can you review this branch in detail?

File: .hooks-config (hunk 1 above)
Comment (aashish24) [inline .hooks-config]: I see. I guess we would have fix it. @dlonie can you do that?

Comment (dlonie) [inline .hooks-config]: Is there a reply missing here? I just see @aashish24 asking a question and then replying "I see" -- what happened in between?

File: Packages/DV3D/ButtonBarWidget.py [hunk 2]
```diff
@@ -58,23 +58,26 @@ def __init__( self, iren, **args ):
         self.buttonWidget.SetInteractor(self.renderWindowInteractor)
         self.buttonWidget.SetRepresentation( self.buttonRepresentation )
         self.buttonWidget.AddObserver( 'StateChangedEvent', self.processStateChangeEvent )
+#        self.buttonWidget.AddObserver( 'AnyEvent', self.processButtonEvent )
```
Comment (dlonie) [inline Packages/DV3D/ButtonBarWidget.py]: This should be taken out completely if not used.

File: Packages/DV3D/ButtonBarWidget.py [hunk 3]
```diff
@@ -58,23 +58,26 @@ def __init__( self, iren, **args ):
...
         self.buttonRepresentation.Highlight( self._state )
         self.buttonWidget.SetPriority(1.0)
         self.updateWidgetState()
 #         if self.id == 'ToggleVolumePlot':
 #             print "."
 
+    def processButtonEvent(self, *args ):
+        print '-----> processButtonEvent: ', str( args )
```
Comment (dlonie) [inline Packages/DV3D/ButtonBarWidget.py]: Leftover debugging output. Should be removed before merging.

File: Packages/DV3D/ButtonBarWidget.py [hunk 4]
```diff
@@ -58,23 +58,26 @@ def __init__( self, iren, **args ):
...
+
     def getState(self):
         return self._state
 
     def setState(self, value):
+#        print "----------------->>> Button [%s] Setting state = %s " % ( self.id, str(value) )
```
Comment (dlonie) [inline Packages/DV3D/ButtonBarWidget.py]: Leftover debugging.

File: Packages/DV3D/ButtonBarWidget.py [hunk 5]
```diff
@@ -140,10 +143,8 @@ def setToggleProps(self, state = None ):
             h_opacity = 0.7 if ( self.getState() == 0 ) else 1.0
             prop.SetOpacity( h_opacity )
             self.buttonRepresentation.SetHoveringProperty(prop)
-            self.buttonRepresentation.Modified()
-            self.buttonRepresentation.NeedToRenderOn()
-            #if self.id == "Volume":
-            #    print "  ---> Set Volume constituent button opacity: ", str( opacity )
+            self.updateWidgetState()
+     #       print "  ---> setToggleProps[%s:%x]: active = %s " % ( self.id, id(self), str( self.getState() ) )
```
Comment (dlonie) [inline Packages/DV3D/ButtonBarWidget.py]: Leftover debugging.

File: Packages/DV3D/ButtonBarWidget.py [hunk 6]
```diff
@@ -162,7 +163,8 @@ def setToggleState( self, state ):
 
     def processStateChangeEvent( self, obj, event, indirect = False ):
         self.invokingEvent = True 
-        self.setButtonState( ( self.getState() + 1 ) % self.numberOfStates )      
+#        print "Button[%s]:processStateChangeEvent(%d)" % ( self.id, self.getState() )
```
Comment (dlonie) [inline Packages/DV3D/ButtonBarWidget.py]: Same

File: Packages/DV3D/ButtonBarWidget.py [hunk 7]
```diff
@@ -233,6 +243,10 @@ def createControlBar( self, name, interactor, build_args, **args ):
             if 'orientation' not in args:   args[ 'orientation' ] = Orientation.Horizontal
             cbar = ControlBar( name, interactor, **args )
             cbar.init( build_args, **args )
+     #       print "Activating *constituent* buttons, cbar '%s':%x" % ( name, id(cbar) )
```
Comment (dlonie) [inline Packages/DV3D/ButtonBarWidget.py]: Same

File: Packages/DV3D/ButtonBarWidget.py [hunk 8]
```diff
@@ -281,12 +295,13 @@ def restoreInteractionState( self, state ):
                                 print>>sys.stderr, "Error, interaction state conflict: %s vs %s " % ( configFunct.name, bbar.InteractionState) 
                                 return
                         bbar.InteractionState = configFunct.name 
-                        #print " ---> Set Plot bbar interaction State: ", configFunct.name
+#                        print " ---> Set Plot bbar interaction State: ", configFunct.name
```
Comment (dlonie) [inline Packages/DV3D/ButtonBarWidget.py]: Let's just take this out instead.

File: Packages/DV3D/ButtonBarWidget.py [hunk 9]
```diff
@@ -442,7 +465,21 @@ def init( self, build_args, **args ):   # ( "Step", ("Run","Stop") ), self.proce
                 self.addButton( bspec[0], toggle=bspec[1], **args ) 
             else: 
                 self.addButton( bspec, **args )
-            
+
+    def changeButtonActivations(self, activation_list ):
+#        print " ** Change Button Activations: ", str( activation_list )
```
Comment (dlonie) [inline Packages/DV3D/ButtonBarWidget.py]: I'll stop spamming notifications on these, but let's make a pass at cleaning up the rest of the commented-out code before merging.

**Comment by aashish24** [comment]
@dlonie these are good comments. But do you see some more critical issues with the branch? After that can you merge locally and run ctest?

**Comment by dlonie** [comment]
I'm still looking, @aashish24. This is a 3k line change.

**Comment by aashish24** [comment]
Sure :smile:  yes it is a big branch. I told @ThomasMaxwell that one debug statement per method or per class (commented out) is Okay.

**Comment by ThomasMaxwell** [comment]
I think time is better spent merging, testing the merged branch, and then patching errors if necessary.   It is very difficult to find errors by looking at code (particularly if you didn't write it).  We have a dedicated tester here at NASA who has been waiting for the branch to be merged so she can begin testing .

-- Tom

**Comment by dlonie** [comment]
Ok, I'll just run the tests and report back. It'd be nice to see some of
the uncommented debug statements removed before merging, but eh. Whatever
the team thinks is best.

**Comment by aashish24** [comment]
@dlonie @ThomasMaxwell I wil send an email to the developers list on what is Ok and what is not in terms of leaving the debug statement. I wouldn't worry about it here. 

@dlonie I would just check on VTK related changes and ctest.

**Comment by dlonie** [comment]
Building now, will +1 after running tests.

I'll go with flow on this (it certainly makes my job easier!), but I can't help but point out that this sort of approach is exactly why VTKPlots needs a complete rewrite after months of development effort.

**Comment by ThomasMaxwell** [comment]
I'll push a fixup branch after the main branch is merged and tested.   In that branch I'll include the removal of any unnecessary/excessive diagnostics.   At this point I think the priority is getting the branch into master so we can thoroughly test it in that context.

**Comment by ThomasMaxwell** [comment]
IMO the problem here is that we are all so busy that the code review of large branches gets pushed back until just before the release deadline, which barely leaves enough time for proper testing of the merged branch.  I would be happy to see the branches reviewed a month before the release date, but that never seems to happen.  As the release deadline gets close I start pushing to get the branch merged, even if that means skipping the code review.

-- Tom

**Comment by dlonie** [comment]
That is indeed a problem.

Also, It's much, much easier to do a thorough review of finer-grained patches -- reviewing a large branch that implements dozens of fixes/additions simultaneously is very difficult. I'm not trying to pick on @ThomasMaxwell here ;) I've seen this happen a number of times on this project from a few developers.

If we stick to one feature / fix per PR, it becomes much easier to spot issues in review, and any issues that are found will only delay that single feature, rather than make months worth of work go on hold.

CTest is running now.

**Comment by dlonie** [comment]
There were a number of test failures: https://www.cdash.org/viewTest.php?onlyfailed&buildid=3706810

The VCS ones I checked are due to the new tick-label format, so they should be ok. The 3D test should be looked at, though -- the vector test looks like it's improved, but there were some segfaults on the others.

**Comment by ThomasMaxwell** [comment]
BTW, as noted in the pull request, there is an associated vistrails branch uvcdat-vcs3D-features_for_v2.1.1.  Testing the uvcdat vcs3D-features_for_v2.1.1  branch with the current vistrails uvcdat-master branch could cause seg faults.   Sorry for not making this more obvious.  I did a local merge with the current uvcdat master and am doing my own testing now.

**Comment by ThomasMaxwell** [comment]
I just ran the ctests on local merges of the vcs3D-features_for_v2.1.1 uvcdat and vistrails branches- all vcs3D tests passed except for dv3d_constituents_test which seg faulted.   Working on fixing that one now.

-- Tom

From: <Maxwell>, "Maxwell, Thomas P. (GSFC-606.2)[SCIENCE APPLICATIONS INTL CORP]" <thomas.maxwell@nasa.gov<mailto:thomas.maxwell@nasa.gov>>
Date: Monday, February 23, 2015 3:11 PM
To: UV-CDAT/uvcdat <reply@reply.github.com<mailto:reply@reply.github.com>>
Subject: Re: [uvcdat] Vcs3D features for v2.1.1 (#951)

BTW, as noted in the pull request, there is an associated vistrails branch uvcdat-vcs3D-features_for_v2.1.1.  Testing the uvcdat vcs3D-features_for_v2.1.1  branch with the current vistrails uvcdat-master branch could cause seg faults.   Sorry for not making this more obvious.  I did a local merge with the current uvcdat master and am doing my own testing now.

**Comment by ThomasMaxwell** [comment]
When I attempt to start uvcdat with my branches merged I get the following error.   Is this a known issue?:

Uncaught exception on initialization: Exception: Error: 'outfill' is not a valid vcs element
Valid vcs elements are: ['projection', 'isofill', 'marker', '3d_dual_scalar', 'texttable', 'template', 'fillarea', 'font', '3d_vector', '1d', 'textorientation', 'textcombined', '3d_scalar', 'xvsy', 'xyvsy', 'isoline', 'boxfill', 'colormap', 'line', 'meshfill', 'yxvsx', 'taylordiagram', 'list', 'fontNumber', 'display', 'vector', 'scatter']
Traceback (most recent call last):
  File "/Developer/Projects/EclipseWorkspace/vistrails/vistrails/uvcdat.py", line 89, in <module>
    v = gui.application.start_application()
  File "/Developer/Projects/EclipseWorkspace/vistrails/vistrails/gui/application.py", line 703, in start_application
    x = VistrailsApplication.init(optionsDict)
  File "/Developer/Projects/EclipseWorkspace/vistrails/vistrails/gui/application.py", line 166, in init
    self.initialize_uvcdat_plots()
  File "/Developer/Projects/EclipseWorkspace/vistrails/vistrails/gui/application.py", line 680, in initialize_uvcdat_plots
    self._plot_manager.load_plots()
  File "/Developer/Projects/EclipseWorkspace/vistrails/vistrails/core/uvcdat/plotmanager.py", line 90, in load_plots
    self.load_vcs_plots()
  File "/Developer/Projects/EclipseWorkspace/vistrails/vistrails/core/uvcdat/plotmanager.py", line 148, in load_vcs_plots
    for m in get_canvas().listelements(str(pl).lower()):
  File "/Developer/Projects/EclipseWorkspace/uvcdat/master/build/install/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/vcs/Canvas.py", line 4439, in listelements
    L = apply(f, args)
  File "/Developer/Projects/EclipseWorkspace/uvcdat/master/build/install/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/vcs/utils.py", line 164, in listelements
    raise Exception,"Error: '%s' is not a valid vcs element\nValid vcs elements are: %s" % (typ,vcs.elements.keys())
Exception: Error: 'outfill' is not a valid vcs element
Valid vcs elements are: ['projection', 'isofill', 'marker', '3d_dual_scalar', 'texttable', 'template', 'fillarea', 'font', '3d_vector', '1d', 'textorientation', 'textcombined', '3d_scalar', 'xvsy', 'xyvsy', 'isoline', 'boxfill', 'colormap', 'line', 'meshfill', 'yxvsx', 'taylordiagram', 'list', 'fontNumber', 'display', 'vector', 'scatter']


From: <Maxwell>, "Maxwell, Thomas P. (GSFC-606.2)[SCIENCE APPLICATIONS INTL CORP]" <thomas.maxwell@nasa.gov<mailto:thomas.maxwell@nasa.gov>>
Date: Monday, February 23, 2015 3:23 PM
To: UV-CDAT/uvcdat <reply@reply.github.com<mailto:reply@reply.github.com>>
Subject: Re: [uvcdat] Vcs3D features for v2.1.1 (#951)

I just ran the ctests on local merges of the vcs3D-features_for_v2.1.1 uvcdat and vistrails branches- all vcs3D tests passed except for dv3d_constituents_test which seg faulted.   Working on fixing that one now.

-- Tom

From: <Maxwell>, "Maxwell, Thomas P. (GSFC-606.2)[SCIENCE APPLICATIONS INTL CORP]" <thomas.maxwell@nasa.gov<mailto:thomas.maxwell@nasa.gov>>
Date: Monday, February 23, 2015 3:11 PM
To: UV-CDAT/uvcdat <reply@reply.github.com<mailto:reply@reply.github.com>>
Subject: Re: [uvcdat] Vcs3D features for v2.1.1 (#951)

BTW, as noted in the pull request, there is an associated vistrails branch uvcdat-vcs3D-features_for_v2.1.1.  Testing the uvcdat vcs3D-features_for_v2.1.1  branch with the current vistrails uvcdat-master branch could cause seg faults.   Sorry for not making this more obvious.  I did a local merge with the current uvcdat master and am doing my own testing now.

**Comment by doutriaux1** [comment]
@ThomasMaxwell I saw this this moring it is fixed at: https://github.com/UV-CDAT/VisTrails/pull/14

**Comment by doutriaux1** [comment]
@ThomasMaxwell I created the VisTrails PR for you at: https://github.com/UV-CDAT/VisTrails/pull/15

In the future please issue the PR at same time and add thelink into a comment so we can see it more easily.

**Comment by ThomasMaxwell** [comment]
It appears the the seg fault in the dv3d_constituents_test is being caused by the VTK upgrade. Were there any changes to the VTK API that I should be accounting for in my code?

From: <Maxwell>, "Maxwell, Thomas P. (GSFC-606.2)[SCIENCE APPLICATIONS INTL CORP]" <thomas.maxwell@nasa.gov<mailto:thomas.maxwell@nasa.gov>>
Date: Monday, February 23, 2015 4:10 PM
To: UV-CDAT/uvcdat <reply@reply.github.com<mailto:reply@reply.github.com>>
Subject: Re: [uvcdat] Vcs3D features for v2.1.1 (#951)

When I attempt to start uvcdat with my branches merged I get the following error.   Is this a known issue?:

Uncaught exception on initialization: Exception: Error: 'outfill' is not a valid vcs element
Valid vcs elements are: ['projection', 'isofill', 'marker', '3d_dual_scalar', 'texttable', 'template', 'fillarea', 'font', '3d_vector', '1d', 'textorientation', 'textcombined', '3d_scalar', 'xvsy', 'xyvsy', 'isoline', 'boxfill', 'colormap', 'line', 'meshfill', 'yxvsx', 'taylordiagram', 'list', 'fontNumber', 'display', 'vector', 'scatter']
Traceback (most recent call last):
  File "/Developer/Projects/EclipseWorkspace/vistrails/vistrails/uvcdat.py", line 89, in <module>
    v = gui.application.start_application()
  File "/Developer/Projects/EclipseWorkspace/vistrails/vistrails/gui/application.py", line 703, in start_application
    x = VistrailsApplication.init(optionsDict)
  File "/Developer/Projects/EclipseWorkspace/vistrails/vistrails/gui/application.py", line 166, in init
    self.initialize_uvcdat_plots()
  File "/Developer/Projects/EclipseWorkspace/vistrails/vistrails/gui/application.py", line 680, in initialize_uvcdat_plots
    self._plot_manager.load_plots()
  File "/Developer/Projects/EclipseWorkspace/vistrails/vistrails/core/uvcdat/plotmanager.py", line 90, in load_plots
    self.load_vcs_plots()
  File "/Developer/Projects/EclipseWorkspace/vistrails/vistrails/core/uvcdat/plotmanager.py", line 148, in load_vcs_plots
    for m in get_canvas().listelements(str(pl).lower()):
  File "/Developer/Projects/EclipseWorkspace/uvcdat/master/build/install/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/vcs/Canvas.py", line 4439, in listelements
    L = apply(f, args)
  File "/Developer/Projects/EclipseWorkspace/uvcdat/master/build/install/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/vcs/utils.py", line 164, in listelements
    raise Exception,"Error: '%s' is not a valid vcs element\nValid vcs elements are: %s" % (typ,vcs.elements.keys())
Exception: Error: 'outfill' is not a valid vcs element
Valid vcs elements are: ['projection', 'isofill', 'marker', '3d_dual_scalar', 'texttable', 'template', 'fillarea', 'font', '3d_vector', '1d', 'textorientation', 'textcombined', '3d_scalar', 'xvsy', 'xyvsy', 'isoline', 'boxfill', 'colormap', 'line', 'meshfill', 'yxvsx', 'taylordiagram', 'list', 'fontNumber', 'display', 'vector', 'scatter']


From: <Maxwell>, "Maxwell, Thomas P. (GSFC-606.2)[SCIENCE APPLICATIONS INTL CORP]" <thomas.maxwell@nasa.gov<mailto:thomas.maxwell@nasa.gov>>
Date: Monday, February 23, 2015 3:23 PM
To: UV-CDAT/uvcdat <reply@reply.github.com<mailto:reply@reply.github.com>>
Subject: Re: [uvcdat] Vcs3D features for v2.1.1 (#951)

I just ran the ctests on local merges of the vcs3D-features_for_v2.1.1 uvcdat and vistrails branches- all vcs3D tests passed except for dv3d_constituents_test which seg faulted.   Working on fixing that one now.

-- Tom

From: <Maxwell>, "Maxwell, Thomas P. (GSFC-606.2)[SCIENCE APPLICATIONS INTL CORP]" <thomas.maxwell@nasa.gov<mailto:thomas.maxwell@nasa.gov>>
Date: Monday, February 23, 2015 3:11 PM
To: UV-CDAT/uvcdat <reply@reply.github.com<mailto:reply@reply.github.com>>
Subject: Re: [uvcdat] Vcs3D features for v2.1.1 (#951)

BTW, as noted in the pull request, there is an associated vistrails branch uvcdat-vcs3D-features_for_v2.1.1.  Testing the uvcdat vcs3D-features_for_v2.1.1  branch with the current vistrails uvcdat-master branch could cause seg faults.   Sorry for not making this more obvious.  I did a local merge with the current uvcdat master and am doing my own testing now.

**Comment by dlonie** [comment]
Is this ready to "rereview"? I have some free cycles I can use to build/test.

**Comment by ThomasMaxwell** [comment]
Yes, the vcs3D_features_for_v2.1.1 branch is now ready for re-review.   Please note that the baseline images have been updated and there is an associated vistrails branch uvcdat-vcs3D-features_for_v2.1.1.

Thanks,
Tom

**Comment by dlonie** [comment]
Is there a pull request for the testdata repository?

**Comment by ThomasMaxwell** [comment]
No PR on the uvcdat-testdata- I pushed directly to master.

**Comment by dlonie** [comment]
Ah...let's try not to do that. It puts the baselines out of sync for everyone else until these changes get merged. See https://github.com/UV-CDAT/uvcdat/wiki/Dashboard,-Testing,-and-Validation#testdata-repository

Testing now.

**Comment by dlonie** [comment]
Still failing tests for me:

https://www.cdash.org/viewTest.php?onlyfailed&buildid=3709751

It looks like most of the issues in vcs are due to difference in master, but there are conflict when I try to merge master in. Please merge master in, resolve the conflicts, and push the resolved branch to this PR.

Outside of VCS, one of the dv3d tests is still failing:

https://www.cdash.org/testDetails.php?test=316072771&build=3709751

Looks like the Valid image is wrong, and the Test image is correct in this case?

**Comment by dlonie** [comment]
Ah, I just remembered #950, which would explain why the "Valid" image has the error.

**Comment by ThomasMaxwell** [comment]
The ctests all passed for me.  I'm guessing this is a merge issue.  My merged branches are: master_with_vcs3D_features_for_v2.1.1 (uvcdat) and uvcdat-master-with-vcs3D-features_for_v2.1.1 (vistrails).

**Comment by dlonie** [comment]
Can we try again with master merged back in? It currently can't be merged with the existing conflicts so we'll need to do this anyway.

**Comment by ThomasMaxwell** [comment]
I've regenerated uvcdat master_with_vcs3D_features_for_v2.1.1 by merging uvcdat master into vcs3D_features_for_v2.1.1.
I've also regenerated vistrails uvcdat-master-with-vcs3D-features_for_v2.1.1 by merging vistrails uvcdat-master into uvcdat-vcs3D-features_for_v2.1.1.
With these new merged branches all the vcs3D ctests pass for me.

-- Tom

**Comment by dlonie** [comment]
Testing the new branches.

**Comment by dlonie** [comment]
Much closer. I still see a couple of failures:

https://www.cdash.org/viewTest.php?onlyfailed&buildid=3711104

`vcs_test_vectors_amplitude` was recently changed by @doutriaux1 in UV-CDAT/uvcdat-testdata@79b06ec38a652673e43012131db867bbb54d9062, so that's probably just a syncing issue with the code and the testdata repo.

`dv3d_vector_test` still has that weird random line across the rendering.

I'd like to see that fixed before merging into master, but I have a feeling I'm in the minority on this. @doutriaux1 @aashish24 feel free to merge if you're ok with this branch.

**Comment by aashish24** [comment]
I am ok with this. one test fail is fixed, the other one seems like a minor issue.

**Comment by ThomasMaxwell** [comment]
Remi,

I'm surprised that you're seeing weird random lines in the vector test.  I've inserted what I see below.   Not seeing random lines.   What OS are you running in?

-- Tom




[cid:ED77F7E2-4F89-41E3-AB54-9ED0CF9956E8]

**Comment by dlonie** [comment]
(This is Dave ;) )

I'm using arch linux.

**Comment by ThomasMaxwell** [comment]
Hi Dave,
    Sorry about the case of mistaken identity.   :-)     If you get a chance it would be helpful to see a screen shot of the weird lines you are encountering in linux.

**Comment by dlonie** [comment]
Here's the link with the images (I posted it on the PR on github, but maybe it didn't go through?)

https://www.cdash.org/viewTest.php?onlyfailed&buildid=3711104

**Comment by dlonie** [comment]
The images are accessed by clicking on the "Failed" text.

**Comment by ThomasMaxwell** [comment]
Thanks Aashish and Dave.   I believe the corresponding vistrails PR still needs to be merged: https://github.com/UV-CDAT/VisTrails/pull/15

-- Tom

**Comment by aashish24** [comment]
@doutriaux1 merged it.

**Pull request merged**